File: benchmarking/cpp_bench/bench_write.cpp
#define MCAP_IMPLEMENTATION
#include "mcap/writer.hpp"

#include <algorithm>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <sys/resource.h>
#include <sys/stat.h>
#include <time.h>
#include <utility>
#include <vector>

/* Shared payload blob parameters; must match gen_blob.py and the other
 * language benches. Message i's payload is the window of the blob starting
 * at (i * kStride) % kWindowSpan, so all implementations feed identical
 * bytes to their writers. */
static const size_t kBlobSize = 16777216;
static const size_t kMaxPayload = 524288;
static const size_t kWindowSpan = kBlobSize - kMaxPayload;
static const uint64_t kStride = 7919;

static size_t payload_offset(uint64_t msg_index)
{
  return static_cast<size_t>((msg_index * kStride) % kWindowSpan);
}

static bool load_blob(const char* path, std::vector<std::byte>& blob)
{
  FILE* f = fopen(path, "rb");
  if (f == nullptr) {
    fprintf(stderr, "Failed to open blob file: %s\n", path);
    return false;
  }
  blob.resize(kBlobSize);
  size_t nread = fread(blob.data(), 1, kBlobSize, f);
  bool at_eof = (fgetc(f) == EOF);
  fclose(f);
  if (nread != kBlobSize || !at_eof) {
    fprintf(stderr, "Blob file %s is not exactly %zu bytes\n", path, kBlobSize);
    return false;
  }
  return true;
}

/* CRC-32 (IEEE, zlib-compatible) over the payload stream, used by
 * run_bench.sh to verify all languages fed identical bytes. */
static uint32_t crc32_update(uint32_t crc, const std::byte* data, size_t len)
{
  static uint32_t table[256];
  static bool table_init = false;
  if (!table_init) {
    for (uint32_t i = 0; i < 256; i++) {
      uint32_t c = i;
      for (int k = 0; k < 8; k++) {
        c = (c & 1) ? 0xEDB88320U ^ (c >> 1) : c >> 1;
      }
      table[i] = c;
    }
    table_init = true;
  }
  crc ^= 0xFFFFFFFFU;
  for (size_t i = 0; i < len; i++) {
    crc = table[(crc ^ static_cast<uint32_t>(data[i])) & 0xFFU] ^ (crc >> 8);
  }
  return crc ^ 0xFFFFFFFFU;
}

/* ru_maxrss is KB on Linux but bytes on macOS; normalize to KB. */
static long peak_rss_kb(void)
{
  struct rusage ru;
  getrusage(RUSAGE_SELF, &ru);
#ifdef __APPLE__
  return ru.ru_maxrss / 1024;
#else
  return ru.ru_maxrss;
#endif
}

int main(int argc, char* argv[])
{
  if (argc != 6) {
    fprintf(stderr, "Usage: %s <output_file> <mode> <num_messages> <payload_size> <blob_file>\n", argv[0]);
    fprintf(stderr, "  mode: unchunked | chunked | zstd | lz4\n");
    return 1;
  }

  const char* filename = argv[1];
  const char* mode = argv[2];
  bool mixed_mode = (strcmp(argv[4], "mixed") == 0);
  long num_messages = mixed_mode ? 0 : strtol(argv[3], nullptr, 10);
  long payload_size = mixed_mode ? 0 : strtol(argv[4], nullptr, 10);

  if (!mixed_mode && (num_messages <= 0 || payload_size <= 0)) {
    fprintf(stderr, "num_messages and payload_size must be positive\n");
    return 1;
  }
  if (!mixed_mode && static_cast<size_t>(payload_size) > kMaxPayload) {
    fprintf(stderr, "payload_size must be <= %zu\n", kMaxPayload);
    return 1;
  }

  std::vector<std::byte> blob;
  if (!load_blob(argv[5], blob)) {
    return 1;
  }

  mcap::McapWriterOptions opts("bench");
  opts.library = "cpp-bench";

  if (strcmp(mode, "unchunked") == 0) {
    opts.noChunking = true;
    opts.compression = mcap::Compression::None;
  } else if (strcmp(mode, "chunked") == 0) {
    opts.chunkSize = 786432;
    opts.compression = mcap::Compression::None;
  } else if (strcmp(mode, "zstd") == 0) {
    opts.chunkSize = 786432;
    opts.compression = mcap::Compression::Zstd;
  } else if (strcmp(mode, "lz4") == 0) {
    opts.chunkSize = 786432;
    opts.compression = mcap::Compression::Lz4;
  } else {
    fprintf(stderr, "Unknown mode: %s\n", mode);
    return 1;
  }

  mcap::McapWriter writer;
  auto res = writer.open(filename, opts);
  if (!res.ok()) {
    fprintf(stderr, "Failed to open writer: %s\n", res.message.c_str());
    return 1;
  }

  if (mixed_mode) {
    /* Mixed payload mode: simulate a 10-second robot recording */

    /* Channel definitions: topic, schema_name, payload_size(s), period_ns, count */
    struct ChannelDef {
      const char* topic;
      const char* schema_name;
      std::vector<size_t> payload_sizes;
      uint64_t period_ns;
      long count;
    };

    ChannelDef channel_defs[] = {
      {"/imu",                "IMU",             {96},                           5000000ULL,   2000},
      {"/odom",               "Odometry",        {296},                         20000000ULL,    500},
      {"/tf",                 "TFMessage",       {80, 160, 320, 800, 1600},    10000000ULL,   1000},
      {"/lidar",              "PointCloud2",     {230400},                     100000000ULL,    100},
      {"/camera/compressed",  "CompressedImage", {524288},                     66666667ULL,    150},
    };
    const int num_channels = 5;

    /* Register schemas and channels (not timed) */
    mcap::Schema schemas[5];
    mcap::Channel channels[5];
    for (int c = 0; c < num_channels; c++) {
      schemas[c] = mcap::Schema(channel_defs[c].schema_name, "jsonschema", "{\"type\":\"object\"}");
      writer.addSchema(schemas[c]);
      channels[c] = mcap::Channel(channel_defs[c].topic, "json", schemas[c].id);
      writer.addChannel(channels[c]);
    }

    /* Pre-generate sorted message schedule: (timestamp, channel_index) */
    struct ScheduleEntry {
      uint64_t timestamp;
      int channel_index;
    };

    std::vector<ScheduleEntry> schedule;
    schedule.reserve(3750);
    for (int c = 0; c < num_channels; c++) {
      for (long i = 0; i < channel_defs[c].count; i++) {
        ScheduleEntry e;
        e.timestamp = static_cast<uint64_t>(i) * channel_defs[c].period_ns;
        e.channel_index = c;
        schedule.push_back(e);
      }
    }
    std::sort(schedule.begin(), schedule.end(), [](const ScheduleEntry& a, const ScheduleEntry& b) {
      if (a.timestamp != b.timestamp) return a.timestamp < b.timestamp;
      return a.channel_index < b.channel_index;
    });

    num_messages = 3750;

    /* Not timed: CRC of the payload stream for cross-language verification */
    uint32_t payload_crc = 0;
    {
      long crc_seq[5] = {0, 0, 0, 0, 0};
      for (size_t i = 0; i < schedule.size(); i++) {
        int c = schedule[i].channel_index;
        const auto& cdef = channel_defs[c];
        size_t psize = cdef.payload_sizes[static_cast<size_t>(crc_seq[c]) % cdef.payload_sizes.size()];
        crc_seq[c]++;
        payload_crc = crc32_update(payload_crc, blob.data() + payload_offset(i), psize);
      }
    }

    /* Time the message-writing loop + close */
    struct timespec t_start, t_end;
    clock_gettime(CLOCK_MONOTONIC, &t_start);

    /* Track per-channel sequence numbers for tf cycling */
    long chan_seq[5] = {0, 0, 0, 0, 0};

    for (size_t i = 0; i < schedule.size(); i++) {
      const auto& entry = schedule[i];
      int c = entry.channel_index;
      const auto& cdef = channel_defs[c];

      /* Determine payload size (cycles for /tf) and blob window */
      size_t psize;
      if (cdef.payload_sizes.size() == 1) {
        psize = cdef.payload_sizes[0];
      } else {
        psize = cdef.payload_sizes[static_cast<size_t>(chan_seq[c]) % cdef.payload_sizes.size()];
      }

      mcap::Message msg;
      msg.channelId = channels[c].id;
      msg.sequence = static_cast<uint32_t>(chan_seq[c]);
      msg.logTime = entry.timestamp;
      msg.publishTime = entry.timestamp;
      msg.data = blob.data() + payload_offset(i);
      msg.dataSize = psize;
      auto wres = writer.write(msg);
      if (!wres.ok()) {
        fprintf(stderr, "Failed to write message %zu: %s\n", i,
                wres.message.c_str());
        writer.close();
        return 1;
      }

      chan_seq[c]++;
    }

    writer.close();

    clock_gettime(CLOCK_MONOTONIC, &t_end);

    struct stat st;
    if (stat(filename, &st) != 0) {
      fprintf(stderr, "Failed to stat file\n");
      return 1;
    }
    long file_size = static_cast<long>(st.st_size);

    long long elapsed_ns = (long long)(t_end.tv_sec - t_start.tv_sec) * 1000000000LL
                          + (long long)(t_end.tv_nsec - t_start.tv_nsec);
    double wall_sec = static_cast<double>(elapsed_ns) / 1e9;

    /* TSV output: op lang mode num_msgs payload_size file_size elapsed_ns wall_sec peak_rss_kb payload_crc32 */
    printf("write\tcpp\t%s\t%ld\t%s\t%ld\t%lld\t%.6f\t%ld\t%u\n",
           mode, num_messages, "mixed", file_size, elapsed_ns, wall_sec, peak_rss_kb(),
           payload_crc);

  } else {
    /* Fixed payload mode (original code path) */

    /* Schema (not timed) */
    mcap::Schema schema("BenchMsg", "jsonschema", "{\"type\":\"object\"}");
    writer.addSchema(schema);

    /* Channel (not timed) */
    mcap::Channel channel("/bench", "json", schema.id);
    writer.addChannel(channel);

    /* Not timed: CRC of the payload stream for cross-language verification */
    uint32_t payload_crc = 0;
    for (long i = 0; i < num_messages; i++) {
      payload_crc = crc32_update(payload_crc, blob.data() + payload_offset(static_cast<uint64_t>(i)),
                                 static_cast<size_t>(payload_size));
    }

    /* Time the message-writing loop + close */
    struct timespec t_start, t_end;
    clock_gettime(CLOCK_MONOTONIC, &t_start);

    for (long i = 0; i < num_messages; i++) {
      mcap::Message msg;
      msg.channelId = channel.id;
      msg.sequence = static_cast<uint32_t>(i);
      msg.logTime = static_cast<uint64_t>(i) * 1000;
      msg.publishTime = msg.logTime;
      msg.data = blob.data() + payload_offset(static_cast<uint64_t>(i));
      msg.dataSize = static_cast<size_t>(payload_size);
      auto wres = writer.write(msg);
      if (!wres.ok()) {
        fprintf(stderr, "Failed to write message %ld: %s\n", i,
                wres.message.c_str());
        writer.close();
        return 1;
      }
    }

    writer.close();

    clock_gettime(CLOCK_MONOTONIC, &t_end);

    struct stat st;
    if (stat(filename, &st) != 0) {
      fprintf(stderr, "Failed to stat file\n");
      return 1;
    }
    long file_size = static_cast<long>(st.st_size);

    long long elapsed_ns = (long long)(t_end.tv_sec - t_start.tv_sec) * 1000000000LL
                          + (long long)(t_end.tv_nsec - t_start.tv_nsec);
    double wall_sec = static_cast<double>(elapsed_ns) / 1e9;

    /* TSV output: op lang mode num_msgs payload_size file_size elapsed_ns wall_sec peak_rss_kb payload_crc32 */
    printf("write\tcpp\t%s\t%ld\t%ld\t%ld\t%lld\t%.6f\t%ld\t%u\n",
           mode, num_messages, payload_size, file_size, elapsed_ns, wall_sec, peak_rss_kb(),
           payload_crc);
  }

  return 0;
}

File: benchmarking/rust_bench/src/bin/bench_write.rs
use mcap::write::WriteOptions;
use mcap::Compression;
use std::collections::BTreeMap;
use std::io::BufWriter;
use std::time::Instant;

/// Shared payload blob parameters; must match gen_blob.py and the other
/// language benches. Message i's payload is the window of the blob starting
/// at (i * STRIDE) % WINDOW_SPAN, so all implementations feed identical
/// bytes to their writers.
const BLOB_SIZE: usize = 16_777_216;
const MAX_PAYLOAD: usize = 524_288;
const WINDOW_SPAN: u64 = (BLOB_SIZE - MAX_PAYLOAD) as u64;
const STRIDE: u64 = 7919;

fn payload_offset(msg_index: u64) -> usize {
    ((msg_index * STRIDE) % WINDOW_SPAN) as usize
}

fn load_blob(path: &str) -> Vec<u8> {
    let blob = std::fs::read(path).expect("Failed to read blob file");
    assert!(
        blob.len() == BLOB_SIZE,
        "Blob file {} is not exactly {} bytes",
        path,
        BLOB_SIZE
    );
    blob
}

/// CRC-32 (IEEE, zlib-compatible) over the payload stream, used by
/// run_bench.sh to verify all languages fed identical bytes.
fn crc32_table() -> [u32; 256] {
    let mut table = [0u32; 256];
    for (i, entry) in table.iter_mut().enumerate() {
        let mut c = i as u32;
        for _ in 0..8 {
            c = if c & 1 != 0 { 0xEDB88320 ^ (c >> 1) } else { c >> 1 };
        }
        *entry = c;
    }
    table
}

fn crc32_update(table: &[u32; 256], crc: u32, data: &[u8]) -> u32 {
    let mut crc = crc ^ 0xFFFFFFFF;
    for &byte in data {
        crc = table[((crc ^ byte as u32) & 0xFF) as usize] ^ (crc >> 8);
    }
    crc ^ 0xFFFFFFFF
}

/// ru_maxrss is KB on Linux but bytes on macOS; normalize to KB.
fn peak_rss_kb() -> libc::c_long {
    let mut rusage: libc::rusage = unsafe { std::mem::zeroed() };
    unsafe { libc::getrusage(libc::RUSAGE_SELF, &mut rusage) };
    if cfg!(target_os = "macos") {
        rusage.ru_maxrss / 1024
    } else {
        rusage.ru_maxrss
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 6 {
        eprintln!("Usage: {} <output_file> <mode> <num_messages> <payload_size|mixed> <blob_file>", args[0]);
        eprintln!("  mode: unchunked | chunked | zstd | lz4");
        std::process::exit(1);
    }

    let blob = load_blob(&args[5]);
    let crc_table = crc32_table();

    let filename = &args[1];
    let mode = &args[2];
    let is_mixed = args[4] == "mixed";

    let opts = match mode.as_str() {
        "unchunked" => WriteOptions::new()
            .use_chunks(false)
            .profile("bench")
            .library("rust-bench"),
        "chunked" => WriteOptions::new()
            .compression(None)
            .chunk_size(Some(786432))
            .profile("bench")
            .library("rust-bench"),
        "zstd" => WriteOptions::new()
            .compression(Some(Compression::Zstd))
            .chunk_size(Some(786432))
            .profile("bench")
            .library("rust-bench"),
        "lz4" => WriteOptions::new()
            .compression(Some(Compression::Lz4))
            .chunk_size(Some(786432))
            .profile("bench")
            .library("rust-bench"),
        _ => {
            eprintln!("Unknown mode: {}", mode);
            std::process::exit(1);
        }
    };

    let file = std::fs::File::create(filename).expect("Failed to create file");
    let buf_writer = BufWriter::new(file);
    let mut writer = opts.create(buf_writer).expect("Failed to create writer");

    let schema_data = b"{\"type\":\"object\"}";
    let metadata = BTreeMap::new();

    if is_mixed {
        // ── Mixed-payload mode: simulate a 10-second robot recording ──

        // Channel definitions: (topic, schema_name, base_payload_sizes, period_ns, count)
        struct ChanDef {
            topic: &'static str,
            schema_name: &'static str,
            payload_sizes: &'static [usize],
            period_ns: u64,
            count: u64,
        }
        let chan_defs: [ChanDef; 5] = [
            ChanDef { topic: "/imu",                schema_name: "IMU",             payload_sizes: &[96],                              period_ns: 5_000_000,   count: 2000 },
            ChanDef { topic: "/odom",               schema_name: "Odometry",        payload_sizes: &[296],                             period_ns: 20_000_000,  count: 500  },
            ChanDef { topic: "/tf",                  schema_name: "TFMessage",       payload_sizes: &[80, 160, 320, 800, 1600],         period_ns: 10_000_000,  count: 1000 },
            ChanDef { topic: "/lidar",              schema_name: "PointCloud2",     payload_sizes: &[230_400],                         period_ns: 100_000_000, count: 100  },
            ChanDef { topic: "/camera/compressed",  schema_name: "CompressedImage", payload_sizes: &[524_288],                         period_ns: 66_666_667,  count: 150  },
        ];

        // Register schemas and channels (not timed)
        let mut channel_ids: Vec<u16> = Vec::new();
        for def in &chan_defs {
            let sid = writer
                .add_schema(def.schema_name, "jsonschema", schema_data)
                .expect("Failed to add schema");
            let cid = writer
                .add_channel(sid, def.topic, "json", &metadata)
                .expect("Failed to add channel");
            channel_ids.push(cid);
        }

        // Pre-generate sorted schedule: (timestamp, channel_index)
        let mut schedule: Vec<(u64, usize)> = Vec::new();
        for (ch_idx, def) in chan_defs.iter().enumerate() {
            for msg_i in 0..def.count {
                let timestamp = msg_i * def.period_ns;
                schedule.push((timestamp, ch_idx));
            }
        }
        // Sort by timestamp; ties broken by channel index (stable sort preserves push order)
        schedule.sort_by(|a, b| a.0.cmp(&b.0).then(a.1.cmp(&b.1)));

        // Not timed: CRC of the payload stream for cross-language verification
        let mut payload_crc: u32 = 0;
        {
            let mut crc_seq: [u32; 5] = [0; 5];
            for (i, &(_, ch_idx)) in schedule.iter().enumerate() {
                let def = &chan_defs[ch_idx];
                let seq = crc_seq[ch_idx];
                crc_seq[ch_idx] += 1;
                let payload_size = def.payload_sizes[seq as usize % def.payload_sizes.len()];
                let off = payload_offset(i as u64);
                payload_crc = crc32_update(&crc_table, payload_crc, &blob[off..off + payload_size]);
            }
        }

        // Per-channel sequence counters and message index (for cycling /tf sizes)
        let mut seq_counters: [u32; 5] = [0; 5];

        // ── Timed: message loop + finish ──
        let start = Instant::now();

        for (i, &(timestamp, ch_idx)) in schedule.iter().enumerate() {
            let def = &chan_defs[ch_idx];
            let seq = seq_counters[ch_idx];
            seq_counters[ch_idx] += 1;

            let payload_size = def.payload_sizes[seq as usize % def.payload_sizes.len()];
            let off = payload_offset(i as u64);
            let payload = &blob[off..off + payload_size];

            writer
                .write_to_known_channel(
                    &mcap::records::MessageHeader {
                        channel_id: channel_ids[ch_idx],
                        sequence: seq,
                        log_time: timestamp,
                        publish_time: timestamp,
                    },
                    payload,
                )
                .expect("Failed to write message");
        }

        writer.finish().expect("Failed to finish");

        let elapsed = start.elapsed();
        let elapsed_ns = elapsed.as_nanos();
        let wall_sec = elapsed.as_secs_f64();

        let file_size = std::fs::metadata(filename)
            .expect("Failed to stat file")
            .len();

        let peak_rss_kb = peak_rss_kb();

        println!(
            "write\trust\t{}\t{}\t{}\t{}\t{}\t{:.6}\t{}\t{}",
            mode, 3750, "mixed", file_size, elapsed_ns, wall_sec, peak_rss_kb, payload_crc
        );
    } else {
        // ── Fixed-payload mode (original behavior) ──
        let num_messages: u64 = args[3].parse().expect("invalid num_messages");
        let payload_size: usize = args[4].parse().expect("invalid payload_size");
        assert!(payload_size <= MAX_PAYLOAD, "payload_size must be <= {}", MAX_PAYLOAD);

        let schema_id = writer
            .add_schema("BenchMsg", "jsonschema", schema_data)
            .expect("Failed to add schema");

        let channel_id = writer
            .add_channel(schema_id, "/bench", "json", &metadata)
            .expect("Failed to add channel");

        // Not timed: CRC of the payload stream for cross-language verification
        let mut payload_crc: u32 = 0;
        for i in 0..num_messages {
            let off = payload_offset(i);
            payload_crc = crc32_update(&crc_table, payload_crc, &blob[off..off + payload_size]);
        }

        // Timed: message loop + finish
        let start = Instant::now();

        for i in 0..num_messages {
            let log_time = i * 1000;
            let off = payload_offset(i);
            writer
                .write_to_known_channel(
                    &mcap::records::MessageHeader {
                        channel_id,
                        sequence: i as u32,
                        log_time,
                        publish_time: log_time,
                    },
                    &blob[off..off + payload_size],
                )
                .expect("Failed to write message");
        }

        writer.finish().expect("Failed to finish");

        let elapsed = start.elapsed();
        let elapsed_ns = elapsed.as_nanos();
        let wall_sec = elapsed.as_secs_f64();

        let file_size = std::fs::metadata(filename)
            .expect("Failed to stat file")
            .len();

        let peak_rss_kb = peak_rss_kb();

        println!(
            "write\trust\t{}\t{}\t{}\t{}\t{}\t{:.6}\t{}\t{}",
            mode, num_messages, payload_size, file_size, elapsed_ns, wall_sec, peak_rss_kb,
            payload_crc
        );
    }
}
